test/utils/padder/padder.go

```diff
@@ -0,0 +1,202 @@
+	}
+
+	padder.Client = cli
+	padder.padRequest = &padRequest{untilAvailableReach: make(corev1.ResourceList)}
+	return padder, nil
+}
+
+// Nodes number of nodes to pad
+func (p *Padder) Nodes(n int) *Padder {
+	p.padRequest.nNodes = n
+	return p
+}
+
+// UntilAvailableReach will pad pods into nodes until reach the desired available untilAvailableReach
+func (p *Padder) UntilAvailableReach(resName corev1.ResourceName, quantity string) *Padder {
+	quan := resource.MustParse(quantity)
+	p.untilAvailableReach[resName] = quan
+	return p
+}
+
+// UntilAvailableReachWithResourceList is like UntilAvailableReach but with a complete ResourceList as a parameter
+func (p *Padder) UntilAvailableReachWithResourceList(resources corev1.ResourceList) *Padder {
+	p.untilAvailableReach = resources
+	return p
+}
+
+// Pad will create pad pods in order to align the nodes
+// with the requested amount of available untilAvailableReach
+func (p *Padder) Pad() error {
+	if p.nNodes == 0 {
+		klog.Warningf("no nodes for padding were found. please specify at least one node")
+		return nil
+	}
+
+	nrtList, err := nrtutil.GetUpdated(p.Client, nrtv1alpha1.NodeResourceTopologyList{}, time.Second*10)
+	if err != nil {
+		return err
+	}
+
+	// create a list of only nodes that have single-numa-node policy
+	singleNumaNrt := nrtutil.FilterTopologyManagerPolicy(nrtList.Items, nrtv1alpha1.SingleNUMANodeContainerLevel)
+	singleNumaNrt = append(singleNumaNrt, nrtutil.FilterTopologyManagerPolicy(nrtList.Items, nrtv1alpha1.SingleNUMANodePodLevel)...)
+
+	for i := 0; i < len(singleNumaNrt) && p.nNodes > 0; i++ {
+		nrt := singleNumaNrt[i]
+		nodePadded := false
+		for _, zone := range nrt.Zones {
+			// check that zone has at least the amount of untilAvailableReach that needed
+			if nrtutil.ZoneResourcesMatchesRequest(zone.Resources, p.untilAvailableReach) {
+				availResList := nrtutil.AvailableFromZone(zone)
+				diffList, err := diffAvailableToExpected(availResList, p.untilAvailableReach)
+				if err != nil {
+					return err
+				}
+
+				// create a pod that asks for exact amount of untilAvailableReach as the diff
+				// in order to reach to desired amount of available untilAvailableReach in the node
+				padPod := objects.NewTestPodPause(p.Namespace, fixture.RandomName("padder"))
+
+				// label the pod with a pad label, so it will be easier to identify later
+				labelPod(padPod, map[string]string{PadderLabel: ""})
+
+				padPod.Spec.Containers[0].Resources.Limits = diffList
+				padPod.Spec.Containers[0].Resources.Requests = diffList
+
+				// place the pod on the selected node
+				//padPod.Spec.NodeSelector = map[string]string{nodes.LabelHostname: nrt.Name}
+				padPod.Spec.NodeSelector = map[string]string{"kubernetes.io/hostname": nrt.Name}
+				if err := p.Client.Create(context.TODO(), padPod); err != nil {
+					return err
+				}
+				klog.InfoS("created pod", "pod", fmt.Sprintf("%s/%s", padPod.Namespace, padPod.Name), "node", nrt.Name)
+				nodePadded = true
```

Comment:
unfortunately not so easy. We need to wait for the pod to go running, to make sure the Kubelet (and the system in general) accpeted it and allocated the resources

Reply:
> We need to wait for the pod to go running
Easy task.
> make sure the Kubelet (and the system in general) accepted it and allocated the resources
Hard. We can't tell in which zone the pod land, hence we can't track the changes under the NRT object.

Reply:
> > We need to wait for the pod to go running
> 
> Easy task.

Let's do in parallel to avoid consuming too much time. And we should likely fail (as in `Expect` failed) if 1+ pods fail to go running. We can reuse some code/ideas from https://github.com/openshift-kni/numaresources-operator/pull/126/commits/41bb6c430cf8be7fa2d65bd8d523419af0d46af2


> > make sure the Kubelet (and the system in general) accepted it and allocated the resources
> 
> Hard. We can't tell in which zone the pod land, hence we can't track the changes under the NRT object.

Aye, hence https://github.com/openshift-kni/numaresources-operator/pull/126

Reply:
relevant snippet I mentioned in the above comment:
```
klog.Infof("e2e infra setup begin")

	mcps, err := machineconfigpools.GetNodeGroupsMCPs(context.TODO(), fxt.Client, nodeGroups)
	Expect(err).ToNot(HaveOccurred())

	klog.Infof("setting e2e infra for %d MCPs", len(mcps))

	var dss []*appsv1.DaemonSet
	for _, mcp := range mcps {
		if mcp.Spec.NodeSelector == nil {
			klog.Warningf("the machine config pool %q does not have node selector", mcp.Name)
			continue
		}

		dsName := objectnames.GetComponentName(numacellmanifests.Prefix, mcp.Name)
		klog.Infof("setting e2e infra for %q: daemonset %q", mcp.Name, dsName)

		ds := numacellmanifests.DaemonSet(mcp.Spec.NodeSelector.MatchLabels, fxt.Namespace.Name, dsName, numaCellDPImage)
		err = fxt.Client.Create(context.TODO(), ds)
		Expect(err).ToNot(HaveOccurred(), "cannot create the numacell daemonset %q in the namespace %q", ds.Name, ds.Namespace)

		dss = append(dss, ds)
	}

	klog.Infof("daemonsets created")

	var wg sync.WaitGroup
	for _, ds := range dss {
		wg.Add(1)
		go func(ds *appsv1.DaemonSet) {
			defer GinkgoRecover()
			defer wg.Done()

			klog.Infof("waiting for daemonset %q to be ready", ds.Name)

			// TODO: what if timeout < period?
			err := e2ewait.ForDaemonSetReady(fxt.Client, ds, 10*time.Second, timeout)
			Expect(err).ToNot(HaveOccurred(), "DaemonSet %q failed to go running")
		}(ds)
	}
	wg.Wait()

	klog.Infof("e2e infra setup completed")
```

Reply:
We can't import the ```gomega``` package for the assertions, for two reasons:
1. we just decided to move padder under ```pkg``` so it's unusual to use to ```gomega``` outside ```test```
2. it's a bad habit to have assertions outside ```ginkgo```'s specs

Reply:
Sure! I'm not against moving into `pkg`, but I wasn't suggesting (let alone requesting) to move it just yet. I'm actually fine both ways.
Also, the snippet above should not be taken verbatim, but rather as illustration as possible flow to make the code more robust.

Reply:
ack. let me revalute

Reply:
@fromanirh Please check the new commit. I hope it address the problem you were concerning about